mysys/lf_hash.c

```diff
@@ -111,13 +111,13 @@ static int l_find(LF_SLIST **head, CHARSET_INFO *cs, uint32 hashnr,
 
     cur_hashnr= cursor->curr->hashnr;
     cur_keylen= cursor->curr->keylen;
-    cur_key= cursor->curr->key;
+    cur_key= my_atomic_loadptr((void *) &cursor->curr->key);
```

Comment:
Hmm... the comment says:
```
    free_ptr_offset     an offset inside the object to a sizeof(void *)
                        memory that is guaranteed to be unused after
                        the object is put in the purgatory. Unused by ANY
                        thread, not only the purgatory owner.
                        This memory will be used to link waiting-to-be-freed
                        objects in a purgatory list.
```

But looking at the code it feels like `LF_SLIST::key` can be used "after the object is put in the purgatory".
E.g.:
```
thr1: finds and pins nodeX
thr2: deletes nodeX and puts it into purgatory
thr1: accesses corrupt key
```

If that's the case, marking it atomic or issuing extra barriers in other places won't help.

Also since `key` must stay constant for the whole node life time, it should naturally be non-atomic.

You could probably use `LF_SLIST::link` for this purpose (although the change ain't going to be trivial), or store local purgatory in some non-intrusive container.

Reply:
Apparently I was reading the code basing on top of a wrong branch, which includes incorrect optimisation for `cur_key`: 2bb4626ab50953b968ca99330c56e984e441ed45. In original code the above situation is impossible, because deleted flag is checked after loading `cur_key`.

So purgatory introduces another release-acquire semantics via `LF_SLIST::key` here. It is unfortunate to have extra barriers in a hot code. But at least you could replace `SEQ_CST` with `ACQUIRE`. And probably fix `LF_SLIST::link` barriers separately.

Reply:
@svoj I assume with 'original code' you refer to the [10.5.9](https://github.com/MariaDB/server/blob/mariadb-10.5.9/mysys/lf_hash.c) or slightly modified, current [10.8](https://github.com/MariaDB/server/blob/10.8/mysys/lf_hash.cc) code base, which does not have patch [2bb4626](https://github.com/MariaDB/server/commit/2bb4626ab50953b968ca99330c56e984e441ed45) applied.

Further, your analysis in the first reply refers to a version having patch [2bb4626](https://github.com/MariaDB/server/commit/2bb4626ab50953b968ca99330c56e984e441ed45) applied.

I also assume that `LF_SLIST::key` can be changed, as it currently is by being used for the purgatory list.



In that case, your initial thoughts on the algorithm having [2bb4626](https://github.com/MariaDB/server/commit/2bb4626ab50953b968ca99330c56e984e441ed45) applied, are correct. This patch introduces an algorithmic bug (not a weak memory bug anymore), which cannot be fixed with barriers.

Without that patch, however, it is necessary to protect `LF_SLIST::key` as you stated. Making the load of `LF_SLIST::key` ACQUIRE is fine. On ARM both memory orderings for load result in the same binary.

Reply:
@mbeck21 right, sorry for confusion. Please disregard my first comment.

Reply:
as far as I remember, purgatory is strictly thread-local and never accessed by other threads. So it should not need any memory barriers, I presume.